linera-core/src/chain_worker/state/mod.rs

```diff
@@ -352,30 +327,51 @@ where
         Ok(())
     }
 
-    /// Returns the blobs requested by their `blob_ids` that are in the chain manager's pending blobs
-    /// and checks that they are otherwise in storage.
-    async fn get_blobs_and_checks_storage(
+    /// Returns the blobs required by the given executed block. The ones that are not passed in
+    /// are read from the chain manager or from storage.
+    async fn get_required_blobs(
         &self,
-        blob_ids: &HashSet<BlobId>,
-    ) -> Result<Vec<Blob>, WorkerError> {
-        let pending_blobs = &self.chain.manager.get().pending_blobs;
-
-        let mut found_blobs = Vec::new();
-        let mut missing_blob_ids = Vec::new();
-        for blob_id in blob_ids {
-            if let Some(blob) = pending_blobs.get(blob_id) {
-                found_blobs.push(blob.clone());
-            } else {
-                missing_blob_ids.push(*blob_id);
+        executed_block: &ExecutedBlock,
+        blobs: &[Blob],
+    ) -> Result<BTreeMap<BlobId, Blob>, WorkerError> {
+        let mut blob_ids = executed_block.required_blob_ids();
+        let manager = self.chain.manager.get();
+
+        let mut found_blobs = BTreeMap::new();
+
+        for blob in manager
+            .proposed
+            .iter()
+            .flat_map(|proposal| &proposal.blobs)
+            .chain(blobs)
+        {
+            if blob_ids.remove(&blob.id()) {
+                found_blobs.insert(blob.id(), blob.clone());
             }
         }
-        let not_found_blob_ids = self.storage.missing_blobs(&missing_blob_ids).await?;
-
-        if not_found_blob_ids.is_empty() {
-            Ok(found_blobs)
-        } else {
-            Err(WorkerError::BlobsNotFound(not_found_blob_ids))
+        blob_ids.retain(|blob_id| {
+            if let Some(blob) = manager.locked_blobs.get(blob_id) {
+                found_blobs.insert(*blob_id, blob.clone());
+                false
+            } else {
+                true
+            }
+        });
+        let missing_blob_ids = blob_ids.into_iter().collect::<Vec<_>>();
+        let blobs_from_storage = self.storage.read_blobs(&missing_blob_ids).await?;
```

Comment:
In this function we look for blobs in:
- `blobs` passed in
- `proposal.blobs`
- `manager.locked_blobs`
- `storage.read_blobs`

is it correct to always look for the blobs in all of these places? I thought that:
- a proposal will have all of its blobs in `proposal.blobs`
- a validated block will have all of its blobs in `locked`
- a proposed block will have blobs in `manager.proposed`
- etc.

Reply:
It's correct to look _anywhere_ because it's content-addressed: If we have the content that matches the blob ID, it's better to use it directly rather than have another node send it to us.

We use this for proposals, validated and confirmed blocks. I'd say:
* For confirmed blocks it makes sense to look up the `locked_blobs`, because it's likely that we've already seen the corresponding validated block and put it in the chain manager. (We could check _that_ first, but I don't think it really speed things up.)
* For validated blocks it makes sense to look up the blobs in the `proposed` block: It's likely that we've seen the proposal first, and put it into the chain manager.
* For proposals maybe we could skip the chain manager step. But it could always be that the new proposal shares blobs with an existing proposal. (Imagine two users trying to publish the same application at the same time, or calling into the same application that will read a certain blob.)

With the current `ChainManager` struct, which for now is still in a single register view, we load either all or none of it from storage, so if we do load it from storage anyway, we might as well look into both `proposed` and `locked_blobs`.

We definitely should improve on this in the future:
* Maybe we can be smarter about where to look, in which case.
* We should make the `ChainManager` a `View` itself, so that we don't load the blobs every time we load the chain manager.

I created https://github.com/linera-io/linera-protocol/issues/3013 and added it to https://github.com/linera-io/linera-protocol/issues/1981.

Reply:
I understand it's _correct_. I just think b/c this method tries to be "callable from anywhere" it's unnecessarily confusing. We should have specific methods for specific case and look only where necessary.

Reply:
- For confirmed blocks we should have all the blobs in the storage, right? 
- For validated blocks it shouldbe in the `manager.locked(_blobs)`?
- For proposals we should expect the blobs are in the proposal and if not - make the client send them. _Our_ clients should be doing that anyway and it will be a long time before anyone writes a 3rd party client.

Reply:
* The blobs used by a confirmed block are getting written to storage, yes. But this is when we handle a _new_ confirmed block certificate, so some of the blobs may not be in our storage _yet_.
* For validated blocks it will be in `manager.locked_blobs` _if_ our own chain manager currently has the same locked block. But depending on the timing and potentially conflicting proposals, we may have no or a different locked block locally.
* This is when we handle a proposal that's possibly new to us, so it generally won't be in our chain manager _yet_. Yes, we do enforce that the client includes the _published_ blobs in the proposal. But the _read_ ones they aren't allowed to send: We are only now validating that proposal and part of validation is to find out whether the blobs already exist, otherwise it's invalid or the client has to prove it to us by sending us a whole _chain_ that published it. But it's true that in this call site they actually _will_ be in storage: We already executed the block just before this. (With https://github.com/linera-io/linera-protocol/issues/2993 this will change slightly but let's worry about that later.)

Reply:
> but let's worry about that later

Or, actually, I'd like to use it as a justification to just all `get_required_blobs` here: In the scenario described in #2993, we are re-proposing a locked block that might read blobs that we don't have in storage. So we _should_ check the `locked_blobs` in our chain manager.